simSPI/linear_simulator/params_utils.py

```diff
@@ -0,0 +1,352 @@
+"""contains functions and classed for parameter generation."""
+from abc import ABCMeta, abstractstaticmethod
+
+import numpy as np
+import starfile
+import torch
+import torch.fft
+from pytorch3d.transforms import (
+    euler_angles_to_matrix,
+    matrix_to_euler_angles,
+    random_rotations,
+)
+
+from .starfile_utils import check_star_file, starfile_opticsparams
+from .utils import U2T
+
+
+def params_update(config):
+    """Update attributes of config.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the dataset generator.
+
+    Returns
+    -------
+    config: class
+    """
+    config.starfile_available = config.input_starfile_path != ""
+    if config.starfile_available:
+        config = starfile_opticsparams(config)
+    config.ctf_size = config.sidelen  # TODO: make it adaptable
+    print(
+        f"Current CTF size has been configured to"
+        f" be equal to the projection size = ({config.sidelen},{config.sidelen})"
+    )
+    return config
+
+
+"""Module to instantiate param generator from a factory of choices."""
+
+
+class ParamsFactory:
+    """Class to instantiate param generator from a factory of choices."""
+
+    def get_params_generator(config):
+        """Return the param generator.
+
+        If the starfile is available then chooses it otherwise
+        chooses distributional param generator.
+
+        Parameters
+        ----------
+        config: class
+
+        Returns
+        -------
+        params_generator: class
+        """
+        if config.starfile_available:
+            return starfile_params(config)
+        else:
+            return distributional_params(config)
+
+
+class Iparams(metaclass=ABCMeta):
+    """Abstract class for the params generator factory with get_params method."""
+
+    @abstractstaticmethod
+    def get_params(self):
+        """Get the parameters."""
+
+
+"""Module to generate parameters using the input star file."""
+
+
+class starfile_params(Iparams):
```

Comment:
for this class, I would modularize things now to have options for relion, cryosparc, etc

You don't have to implement parsing input data, but I would just do the software design so that we can go directly from external programs to compSPI, instead of something like cryosparc -> relion -> compSPI

Reply:
Maybe worth an issue to discuss?

Reply:
Yeah, we had an extensive discussion on this with @fredericpoitevin  and @ynashed. We concluded that we will keep a [factory method pattern](https://www.geeksforgeeks.org/factory-method-python-design-patterns/) for the [params_generator](https://github.com/harshit-gupta-cor/simSPI/blob/0a3601c44fc9e8a72e058dfba74f22231e316398/simSPI/linear_simulator/dataset_generator.py#L34) class. Basically the rest of the code is blind to the params_generator and only tells it where it wants the params to come from (relion, distribution, or cryosparc in future). Each of the different options are put in the [factory](https://github.com/harshit-gupta-cor/simSPI/blob/0a3601c44fc9e8a72e058dfba74f22231e316398/simSPI/linear_simulator/params_utils.py#L44).

Reply:
I put in a comment in the issue on simulator inputs and outputs. We can continue on with the discussion in the issue, even after the PR is merged https://github.com/compSPI/simSPI/issues/14